### PR TITLE
Update/ipv4 drivers

### PR DESCRIPTION
**Pull request by LeStarch**
| | |
|:---|:---|
|**_Originating Project/Creator_**| Infrastrcture |
|**_Affected Component_**|  IP |
|**_Affected Architectures(s)_**|  |
|**_Related Issue(s)_**|  |
|**_Has Unit Tests (y/n)_**| y |
|**_Builds Without Errors (y/n)_**|  y |
|**_Unit Tests Pass (y/n)_**| y |
|**_Documentation Included (y/n)_**|y  |

---
## Change Description

Rework IPv4 code

See #354

### CODE REVIEW

**Comment by LeStarch** [comment]
@jsoref we are seeing some false-positives from the spelling check.  How should I handle these?

**Comment by jsoref** [comment]
@LeStarch: run the shell code in the `To accept these changes, run the following commands from this repository on this branch` (you'll probably need to `git add -u` and then `git commit` or `git commit --amend`)

**Comment by jsoref** [comment]
For `IPv`, you could add:

```
# <word>v#
[Vv]\d+(?:\b|(?=[a-zA-Z_]))
```
to https://github.com/nasa/fprime/blob/devel/.github/actions/spelling/patterns.txt (I'd put it at the very end, order matters).

That'd cause it to be treated as `IP` which should already be accepted.

**Comment by jsoref** [comment]
Notably, that example block comes from https://github.com/check-spelling/check-spelling/wiki/Configuration-Examples:-patterns

**Comment by LeStarch** [comment]
@saba-ja if you get a chance could you review this.  Also, if you have the time, could you run the UTs defined here.

**Review by Josegto1 — APPROVED** [review]

File: Drv/Ip/TcpServerSocket.cpp [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    #include <fioLib.h>
+    #include <hostLib.h>
+    #include <ioLib.h>
+    #include <vxWorks.h>
+    #include <sockLib.h>
+    #include <fioLib.h>
+    #include <taskLib.h>
+    #include <sysLib.h>
+    #include <errnoLib.h>
+    #include <string.h>
```
Comment (saba-ja) [inline Drv/Ip/TcpServerSocket.cpp:29]: Redundant include of fioLib.h on lines 20 and 25

File: Drv/Ip/TcpServerSocket.cpp [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+SocketIpStatus TcpServerSocket::openProtocol(NATIVE_INT_TYPE& fd) {
+    NATIVE_INT_TYPE clientFd = -1;
+    // TCP requires accepting on a the socket to get the client socket file descriptor.
+    if ((clientFd = ::accept(m_base_fd, NULL, NULL)) < 0) {
+        return SOCK_FAILED_TO_ACCEPT; // What we have here is a failure to communicate
+    }
+    // Setup client send timeouts
+    if (IpSocket::setupTimeouts(clientFd) != SOCK_SUCCESS) {
+        ::close(clientFd);
```
Comment (saba-ja) [inline Drv/Ip/TcpServerSocket.cpp:97]: Double Close
socketFd was closed at IpSocket.cpp:71 and is closed again here by close().
Might need to check for the close before attempting to close again.

File: Drv/Ip/UdpSocket.cpp [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    #include <fioLib.h>
+    #include <hostLib.h>
+    #include <ioLib.h>
+    #include <vxWorks.h>
+    #include <sockLib.h>
+    #include <fioLib.h>
+    #include <taskLib.h>
+    #include <sysLib.h>
+    #include <errnoLib.h>
+    #include <string.h>
```
Comment (saba-ja) [inline Drv/Ip/UdpSocket.cpp:30]: Redundant include of fioLib.h on lines 21 and 26

File: Drv/Ip/UdpSocket.cpp [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+        // First IP address to socket sin_addr
+        if ((status = IpSocket::addressToIp4(m_hostname, &(address.sin_addr))) != SOCK_SUCCESS) {
+            ::close(socketFd);
+            return status;
+        };
+
+        // Now apply timeouts
+        if ((status = IpSocket::setupTimeouts(socketFd)) != SOCK_SUCCESS) {
+            ::close(socketFd);
```
Comment (saba-ja) [inline Drv/Ip/UdpSocket.cpp:132]: Double Close
socketFd was closed at IpSocket.cpp:71 and is closed again here by close().
Might need to check for the close before attempting to close again.

File: Drv/TcpServer/test/ut/Tester.cpp [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+            EXPECT_EQ(status2, Drv::SOCK_SUCCESS);
+            EXPECT_EQ(size, m_data_buffer.getSize());
+            Drv::Test::validate_random_buffer(m_data_buffer, buffer);
+
+            // If receive thread is live, try the other way
+            if (recv_thread) {
+                m_spinner = false;
+                m_data_buffer.setSize(sizeof(m_data_storage));
+                client.send(m_data_buffer.getData(), m_data_buffer.getSize());
+                while (not m_spinner) {}
```
Comment (saba-ja) [inline Drv/TcpServer/test/ut/Tester.cpp:88]: I the empty while loop body intentional here?

**Comment by Adude11** [comment]
I was testing some of these changes and noticed the definition of `Drv::SocketIpStatus` in the new `Drv/Ip/IpSocket.hpp` file conflicts with the definition in the old `Drv/SocketIpDriver/SocketIpDriverTypes.hpp`. I can't find anywhere where this has been discussed, so I don't know if this is intentional to force moving to the new modules or not?

**Comment by Joshua-Anderson** [comment]
@Adude11 The plan is to have these new standalone TCP and UDP drivers replace the old SocketIP driver that currently exists. The SocketIP driver was a TCP/UDP combo driver that was really designed specifically for the ground interface's needs. We broke it up into a separate TCP and UDP drivers so that they could be used more easily by other F' components.

File: Drv/CMakeLists.txt [hunk 6]
```diff
@@ -14,4 +14,8 @@ add_fprime_subdirectory("${CMAKE_CURRENT_LIST_DIR}/LinuxGpioDriver/")
 add_fprime_subdirectory("${CMAKE_CURRENT_LIST_DIR}/LinuxSerialDriver/")
 add_fprime_subdirectory("${CMAKE_CURRENT_LIST_DIR}/LinuxSpiDriver/")
 add_fprime_subdirectory("${CMAKE_CURRENT_LIST_DIR}/LinuxI2cDriver/")
+add_fprime_subdirectory("${CMAKE_CURRENT_LIST_DIR}/Ip/")
+add_fprime_subdirectory("${CMAKE_CURRENT_LIST_DIR}/TcpClient/")
+add_fprime_subdirectory("${CMAKE_CURRENT_LIST_DIR}/TcpServer/")
+add_fprime_subdirectory("${CMAKE_CURRENT_LIST_DIR}/Udp/")
 add_fprime_subdirectory("${CMAKE_CURRENT_LIST_DIR}/SocketIpDriver/")
```
Comment (saba-ja) [inline Drv/CMakeLists.txt:21]: `SocketIpDriver` should not be needed in the Drv/cMakeLists.txt after upgrading to new ipv4 drivers.

File: Drv/Ip/IpSocket.cpp [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+SocketIpStatus IpSocket::send(const U8* const data, const U32 size) {
+    U32 total = 0;
+    I32 sent  = 0;
+    U32 i = 0;
+    // Prevent transmission before connection, or after a disconnect
+    if (this->m_fd == -1) {
+        return SOCK_DISCONNECTED;
+    }
+    // Attempt to send out data and retry as necessary
+    for (i = 0; (i < SOCKET_MAX_ITERATIONS) && (total < size); i++) {
```
[the remark this answers is not in the document]
Comment (LeStarch) [reply Drv/Ip/IpSocket.cpp:141]: It is used on line 162 as a check for max iterations.

File: Drv/Ip/IpSocket.cpp [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+SocketIpStatus IpSocket::recv(U8* data, I32& req_read) {
+    I32 size = 0;
+    U32 i = 0;
+    // Check for previously disconnected socket
+    if (m_fd == -1) {
+        return SOCK_DISCONNECTED;
+    }
+
+    // Try to read until we fail to receive data
+    for (i = 0; (i < SOCKET_MAX_ITERATIONS) && (size <= 0); i++) {
```
[the remark this answers is not in the document]
Comment (LeStarch) [reply Drv/Ip/IpSocket.cpp:178]: This one I'll fix.

File: Drv/Ip/IpSocket.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (LeStarch) [reply Drv/Ip/IpSocket.cpp:141]: That check is unnecessary.  I'll fix this too.

File: Drv/Ip/test/ut/PortSelector.cpp [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    };
+
+    // When we are setting up for receiving as well, then we must bind to a port
+    if (::bind(socketFd, reinterpret_cast<struct sockaddr *>(&address), sizeof(address)) == -1) {
+        ::close(socketFd);
+        return 0;
+    }
+    socklen_t size = sizeof(address);
+    if (::getsockname(socketFd, ((struct sockaddr *) &address), &size) == -1) {
+        return 0;
```
[the remark this answers is not in the document]
Comment (LeStarch) [reply Drv/Ip/test/ut/PortSelector.cpp:38]: Good catch!

File: Drv/Ip/SocketReadTask.cpp [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+            Fw::Logger::logMsg("[WARNING] Failed to open port with status %d and errno %d\n", status, errno);
+            Os::Task::delay(SOCKET_RETRY_INTERVAL_MS);
+        }
+
+        // If the network connection is open, read from it
+        if (self->getSocketHandler().isOpened() and (not self->m_stop)) {
+            Fw::Buffer buffer = self->getBuffer();
+            U8* data = buffer.getData();
+            FW_ASSERT(data);
+            I32 size = buffer.getSize();
```
[the remark this answers is not in the document]
Comment (LeStarch) [reply Drv/Ip/SocketReadTask.cpp:75]: I'll fix the edge case.

File: Drv/Ip/SocketReadTask.cpp [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+            status = self->getSocketHandler().recv(data, size);
+            if ((status != SOCK_SUCCESS) && (status != SOCK_INTERRUPTED_TRY_AGAIN)) {
+                Fw::Logger::logMsg("[WARNING] Failed to recv from port with status %d and errno %d\n", status, errno);
+                self->getSocketHandler().close();
+                buffer.setSize(0);
+            } else {
+                // Send out received data
+                buffer.setSize(size);
```
[the remark this answers is not in the document]
Comment (LeStarch) [reply Drv/Ip/SocketReadTask.cpp:83]: I'll fix the edge case

File: Drv/TcpServer/test/ut/Tester.cpp (hunk 5 above)
Comment (LeStarch) [reply Drv/TcpServer/test/ut/Tester.cpp:88]: Yes, it just loops until the variable in another thread flips.

File: Drv/Ip/TcpClientSocket.cpp [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+#endif
+
+    // First IP address to socket sin_addr
+    if (IpSocket::addressToIp4(m_hostname, &(address.sin_addr)) != SOCK_SUCCESS) {
+        ::close(socketFd);
+        return SOCK_INVALID_IP_ADDRESS;
+    };
+
+    // Now apply timeouts
+    if (IpSocket::setupTimeouts(socketFd) != SOCK_SUCCESS) {
```
[the remark this answers is not in the document]
Comment (LeStarch) [reply Drv/Ip/TcpClientSocket.cpp:69]: Will remove the line IpSocket.cpp:71 to look like other (library) function calls, which do not close socket for you.

File: Drv/Ip/TcpServerSocket.cpp (hunk 2 above)
Comment (LeStarch) [reply Drv/Ip/TcpServerSocket.cpp:97]: Will remove the line IpSocket.cpp:71 to look like other (library) function calls, which do not close socket for you.

File: Drv/Ip/UdpSocket.cpp (hunk 4 above)
Comment (LeStarch) [reply Drv/Ip/UdpSocket.cpp:132]: Will remove the line IpSocket.cpp:71 to look like other (library) function calls, which do not close socket for you.

File: Drv/Ip/IpSocket.cpp [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+    m_lock.lock();
+    is_open = m_open;
+    m_lock.unLock();
+    return is_open;
+}
+
+void IpSocket::close(void) {
+    m_lock.lock();
+    (void)::shutdown(this->m_fd, SHUT_RDWR);
+    (void)::close(this->m_fd);
```
[the remark this answers is not in the document]
Comment (LeStarch) [reply Drv/Ip/IpSocket.cpp:108]: I was just going to add an if-check to prevent running code when the items are -1.

**Comment by LeStarch** [comment]
@saba-ja, ready for a re-review!

File: config/IpCfg.hpp [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+
+#ifndef REF_IPCFG_HPP
+#define REF_IPCFG_HPP
+
+enum IpCfg {
+    SOCKET_SEND_TIMEOUT_SECONDS = 1,       // Seconds component of timeout
+    SOCKET_SEND_TIMEOUT_MICROSECONDS = 0,  // Milliseconds component of timeout
+    SOCKET_IP_SEND_FLAGS = 0,                 // send, sendto FLAGS argument
+    SOCKET_IP_RECV_FLAGS = 0,                 // recv FLAGS argument
+    SOCKET_MAX_ITERATIONS = 0xFFFF,        // Maximum retry iterations
```
Comment (Joshua-Anderson) [inline config/IpCfg.hpp:21]: Do you think it might make sense to make this is a private constant, not a configuration option?

We don't set recv timeouts, so the only thing that can trigger a retry are interrupts, which we likely want to transparently retry.

While we do have a send timeout, those can't safely be re-tried internally and should likely be surfaced to the user, leaving the only thing that can safely require a retry an interrupt.

I don't think most users will want to tweak the number of times we retry after a signal interrupt, can we reduce the "user" configuration complexity by hiding this constant?

File: config/IpCfg.hpp [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+// ALL RIGHTS RESERVED.  United States Government Sponsorship
+// acknowledged.
+//
+// ======================================================================
+
+#ifndef REF_IPCFG_HPP
+#define REF_IPCFG_HPP
+
+enum IpCfg {
+    SOCKET_SEND_TIMEOUT_SECONDS = 1,       // Seconds component of timeout
```
Comment (Joshua-Anderson) [inline config/IpCfg.hpp:17]: Could we specify that this is the send only timeout?

**Review by Joshua-Anderson — APPROVED** [review]

File: config/IpCfg.hpp (hunk 15 above)
Comment (LeStarch) [reply config/IpCfg.hpp:17]: The name has "SEND" in it.  Is that what you are referring to?

File: Drv/Ip/IpSocket.cpp [hunk 16]
```diff
@@ -0,0 +1,204 @@
...
+    if (this->m_fd == -1) {
+        return SOCK_DISCONNECTED;
+    }
+    // Attempt to send out data and retry as necessary
+    for (U32 i = 0; (i < SOCKET_MAX_ITERATIONS) && (total < size); i++) {
+        sent = 0;
+        // Send using my specific protocol
+        sent = this->sendProtocol(data + total, size - total);
+        // Error is EINTR or timeout just try again
+        if (((sent == -1) && (errno == EINTR)) || (sent == 0)) {
```
[the remark this answers is not in the document]
Comment (LeStarch) [reply Drv/Ip/IpSocket.cpp:146]: Don't think this is possible.  If the `send` variable is zero, it will skip the loop entirely and won't check that condition.  Thus if a user sends 0 bytes, no send call will be triggered.

File: config/IpCfg.hpp (hunk 15 above)
Comment (Joshua-Anderson) [reply config/IpCfg.hpp:17]: Ah, I missed that, I was mostly looking at the comment. Looks good!

File: Drv/Ip/IpSocket.cpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (LeStarch) [reply Drv/Ip/IpSocket.cpp:146]: I don't understand your comment about the timeout.  In the case of a partial send, we just retry the next bytes.  In the case of a timeout on the zeroth byte we retry again.  This will work until we hit the maximum retries....and then we report to the user.

[the remark this answers is not in the document]
Comment (Joshua-Anderson) [reply Drv/Ip/IpSocket.cpp:146]: > Don't think this is possible. If the send variable is zero, it will skip the loop entirely and won't check that condition. Thus if a user sends 0 bytes, no send call will be triggered.

Looking at the loop condition I agree. My concern was getting stuck in the loop if the user provided a send size of zero, but since we never meet the conditions to enter the loop if the size is 0 that isn't a concern.

> I don't understand your comment about the timeout. In the case of a partial send, we just retry the next bytes. In the case of a timeout on the zeroth byte we retry again. This will work until we hit the maximum retries....and then we report to the user.

Ok, with your understanding I now agree the code is correctly handling the timeout case.

[the remark this answers is not in the document]
Comment (Joshua-Anderson) [reply Drv/Ip/IpSocket.cpp:146]: Thinking about it more however, do we really want to transparently handle timeouts? If a user is applying a timeout, my suspicion is that they could be concerned about meeting timing deadlines. If I set a send timeout, I would expect the send call to return within the timeout window to make sure I don't miss other deadlines.

Users could resolve this by dropping down the number of retries, but tweaking this number also affects the number of EINTR signal interrupts that could be received by both send and receive.

Also the defaults as currently set makes it really easy to shoot yourself in the foot. With `0xFF` retries you effectively have a near infinite timeout.

I think I understand your goal of trying to transparently handle timeouts for users, which is useful, I'm just not sure that's what all users setting a timeout want. Some users may want to re-try a timeout, others may want to just treat it as an error and move on.

[the remark this answers is not in the document]
Comment (LeStarch) [reply Drv/Ip/IpSocket.cpp:146]: If the user wants the call to timeout and return within a specific interval, they may simply set the MAX_ITERATIONS configuration to 1 and it will return a `SOCK_INTERRUPTED_TRY_AGAIN`.   

The default paradigm is meant to represent the case of "`send` will send all data and will block until doing so" without the risk of a literal block forever scenario.  This parallels the `recv` functionality which **does** block forever, but is protected on its own thread.  `send` does not have a thread and thus should have a finite limit.

File: config/IpCfg.hpp (hunk 14 above)
Comment (LeStarch) [reply config/IpCfg.hpp:21]: Given we let users set timeouts, we should also allow them to set the max iterations.  I am cleaning up the descriptive comments here to make it clearer what each does.

File: Drv/Ip/IpSocket.cpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (LeStarch) [reply Drv/Ip/IpSocket.cpp:146]: In re-reviewing this, I think I see your point, which is that interrupts and send timeouts are coupled together when they likely should not be.  Send timeouts and retries in the case of interruptions should be handled differently.  I'll see if there is a clean way to pull these apart.

[the remark this answers is not in the document]
Comment (Joshua-Anderson) [reply Drv/Ip/IpSocket.cpp:146]: I'm seeing a few quirks with the current timeout design and I just want to make sure we're on the same page.
Do you think are worth resolving?

- EINTR and TIMEOUT retries are tied together, so if you're using signals you can't safely drop the timeout retries to zero.
- SOCK_INTERRUPTED_TRY_AGAIN has no way of returning how much data was actually sent, so that makes it difficult to handle partial sends after a timeout.
- Users that care about timeouts need to know to set both the timeout and the retry count, otherwise the high retry count will cause send to still effectively block. This seems easy to miss and could be painful to debug.

[the remark this answers is not in the document]
Comment (LeStarch) [reply Drv/Ip/IpSocket.cpp:146]: This code supports specifically the F´components that wrap these items.  Thus we don't have to support "whatever users might want" but more narrowly "what our supplied components expect".    Given this and the fact that the tcp driver is inline with other drivers I'd recommend leaving the code as is.

Some notes:
1. `send timeouts` are reset if a byte is sent, thus the "total time" is inherently hard to predict.
2. Retries are the standard mechanism for dealing with a message that does not send, so iterating a large fixed time is expected.
3. Other drivers in the driver package exhibit this same block until done behavior based on the assumption that `send` will usually succeed.

I'd recommend breaking out clean driver timeouts into its own issue that we can revisit at a later time as it will involve refactoring all of our Linux driver code.  There is a way to implement this correctly, but it is more complicated than should be included here.

[the remark this answers is not in the document]
Comment (Joshua-Anderson) [reply Drv/Ip/IpSocket.cpp:146]: > This code supports specifically the F´components that wrap these items. Thus we don't have to support "whatever users might want" but more narrowly "what our supplied components expect".

Yes, but if don't expose this functionality to the TCP and UDP components users don't have access to this functionality either. 😄 

As you've said, cleanly handling timeouts isn't an easy problem to solve. If `SO_SNDTIMEO` doesn't provide a guaranteed upper time bound it doesn't seem like there's a lot of benefit for F' for properly handling them either.

I'm happy to punt this problem for later if you're okay with it, I was just making a big deal out of it now since it seems like it might require interface changes which will be more difficult to make later on.

[the remark this answers is not in the document]
Comment (LeStarch) [reply Drv/Ip/IpSocket.cpp:146]: The drivers have an enumeration that lists error codes. It would be fairly straight forward to update that to add a TIMEOUT error code without the changing the API.  Thus, I am not concerned with splitting this work out into a new issue for further down the road.

File: Drv/Udp/docs/sdd.md [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+
+bool constructApp(bool dump, U32 port_number, char* hostname) {
+    ...
+    comm.init(0);
+    ...
+    if (hostname != NULL && port_number != 0) {
+        Fw::EightyCharString name("ReceiveTask");
+        comm.configureSend(hostname, port_number);
+        comm.configureRecv(hostname, port_number);
+        comm.startSocketTask(name, TASK_PRIORITY, TASK_STACK_SIZE);
```
Comment (Joshua-Anderson) [inline Drv/Udp/docs/sdd.md:60]: We only need to start the socket task when running in receive mode, correct? We may want to document this and possibly provide an example of a UDP source-only configuration.

**Review by Joshua-Anderson — COMMENTED** [review]
This is looking good, all I had was a couple minor comments!

**Review by Joshua-Anderson — APPROVED** [review]
:shipit:

**Comment by LeStarch** [comment]
Going to merge in the IPv4 drivers.  Next step is to migrate to using these and remove the old SocketIpDriver!
